railties/lib/rails/api/generator.rb

```diff
@@ -0,0 +1,47 @@
+require "sdoc"
+
+class RDoc::Generator::API < RDoc::Generator::SDoc # :nodoc:
+  RDoc::RDoc.add_generator self
+
+  alias :original_generate_class_tree_level :generate_class_tree_level
+
+  def generate_class_tree_level(classes, visited = {})
+    # If the method is called for the first time, we just
+    # extract the core classes from the list.
+    first_call = visited == {}
+
+    if first_call
+      core_exts = []
+
+      classes.delete_if do |klass|
+        core_exts << klass if !internal_class?(klass.full_name)
+      end
+    end
+
+    tree = original_generate_class_tree_level(classes, visited)
+
+    if first_call
+      core_ext_item = ["Core extensions", "", "", []]
+
+      core_exts.each do |klass|
+        core_ext_item[3] << [
+          klass.name,
+          klass.document_self_or_methods ? klass.path : "",
+          "",
+          []
+        ]
+      end
+
+      tree.unshift(core_ext_item)
+    end
+
+    tree
+  end
+
+  def internal_class?(name)
```

Comment:
Think the `internal_class` naming is weird since internal makes one think of private, as in private API. How about `def public_constant?` for the method name?

Alternatively, I'd do:

```ruby
PUBLIC_CONSTANTS = %w( AbstractController ActionCable ActionController … )

# Elsewhere…
PUBLIC_CONSTANTS.include?(klass.full_name)
```